lib/WUI/nhttp/file_info.cpp

```diff
@@ -23,19 +23,116 @@ FileInfo::FileInfo(const char *filename, bool can_keep_alive, bool after_upload)
     : can_keep_alive(can_keep_alive)
     , after_upload(after_upload) {
     strlcpy(this->filename, filename, sizeof this->filename);
+    /*
+     * Eat the last slash on directories.
+     *
+     * This is mostly aestetic only... otherwise we get double slashs in the
+     * results/URLs. That works, but, meh...
+     */
+    const size_t len = strlen(this->filename);
+    if (len > 0 && this->filename[len - 1] == '/') {
+        this->filename[len - 1] = '\0';
+    }
 }
 
 Step FileInfo::step(std::string_view, bool, uint8_t *output, size_t output_size) {
-    // TODO: Support directories.
-
-    // We assume the answer fits. If it doesn't, it doesn't overwrite anything, it just truncates the answer.
-    // We'll have to revisit this for directories, of course.
+    /*
+     * We assume the answer fits. If it doesn't, it doesn't overwrite anything, it just truncates the answer.
+     * Directories send each file as a single packet, but that one also assumes it fits.
+     */
     ConnectionHandling handling = can_keep_alive ? ConnectionHandling::ChunkedKeep : ConnectionHandling::Close;
     struct stat finfo;
-    if (stat(filename, &finfo) == 0) {
-        if (finfo.st_mode & S_IFDIR) {
-            return Step { 0, 0, StatusPage(Status::NotImplemented, can_keep_alive, "Directory listing not implemented") };
-        } else {
+    const size_t last_chunk_len = strlen(LAST_CHUNK);
+
+    if (dir) {
+        /*
+         * We already have a directory listing running. Try to pull another file out of there.
+         */
+        struct dirent *ent;
+        while ((ent = readdir(dir.get()))) {
+            // Check this is GCODE, not some other thing. Skip the others.
+            const char *last_dot = strrchr(ent->d_name, '.');
+            if (last_dot == nullptr) {
+                continue;
+            }
+            if (strcasecmp(last_dot, ".gcode") != 0 && strcasecmp(last_dot, ".gco") != 0) {
```

Comment:
there are may-be some other supported suffices as well ... `.g`? Ideally if we could unify this with the lazy file list impl.

Reply:
oh, it has moved to `src/gui/file_raii.hpp`
`static bool FnameMatchesPattern(const char *fname)`

probably already usable...